iOS Client/ArticlesController/MBArticlesViewController.swift

```diff
@@ -109,73 +103,84 @@ class MBArticlesViewController: UIViewController, UITableViewDelegate, UITableVi
+                let topLevelCategories = Set(categories.flatMap({ (category) -> String? in
+                    return (category as? MBCategory)?.getTopLevelCategory()?.name
+                }))
+            
+                for topLevelCategory in topLevelCategories {
+                    if let _ = retVal[topLevelCategory] {
+                        retVal[topLevelCategory]?.append(article)
+                    } else {
+                        retVal[topLevelCategory] = [article]
+                    }
+                }
+                
+                if topLevelCategories.count == 0 {
+                    print("Excluding article! \(article.title ?? "<no title>") b/c it has no categories")
+                }
+            } else {
+                print("Excluding article \(article.title ?? "<no title>") b/c it has no categories")
             }
         }
+        return retVal
     }
     
     override func didReceiveMemoryWarning() {
         super.didReceiveMemoryWarning()
         // Dispose of any resources that can be recreated.
     }
-    
-    // MARK: - showTimestamps() is used for debug purposes
-    @objc func showTimestamps() {
-        let lastOverallUpdate = UserDefaults.standard.double(forKey: MBConstants.DEFAULTS_KEY_ARTICLE_UPDATE_TIMESTAMP)
-        let lastBackgroundUpdate = UserDefaults.standard.double(forKey: MBConstants.DEFAULTS_KEY_BACKGROUND_APP_REFRESH_TIMESTAMP)
-        let lastBackgroundAttempt = UserDefaults.standard.double(forKey: MBConstants.DEFAULTS_KEY_BACKGROUND_APP_REFRESH_ATTEMPT_TIMESTAMP)
-        
-        let lastOverallUpdateDate = Date(timeIntervalSinceReferenceDate: lastOverallUpdate)
-        let lastBackgroundUpdateDate = Date(timeIntervalSinceReferenceDate: lastBackgroundUpdate)
-        let lastBackgroundAttemptDate = Date(timeIntervalSinceReferenceDate: lastBackgroundAttempt)
-        
-        let dateFormatter = DateFormatter()
-        if let timeZone = TimeZone(identifier: "America/New_York") {
-            dateFormatter.timeZone = timeZone
-        }
-        dateFormatter.locale = NSLocale.current
-        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm"
-        
-        let msg = "Last Update: \(dateFormatter.string(from: lastOverallUpdateDate))\nLast Background Refresh: \(dateFormatter.string(from: lastBackgroundUpdateDate))\nLast Background Attempt: \(dateFormatter.string(from: lastBackgroundAttemptDate))"
-        
-        let ac = UIAlertController(title: "DEBUG", message: msg, preferredStyle: .actionSheet)
-        ac.addAction(UIAlertAction(title: "👍", style: .default, handler: nil))
-        self.present(ac, animated: true, completion: nil)
-    }
 }
 
-// MARK: - UITableViewDataSource
 extension MBArticlesViewController {
+    // MARK: - UITableViewDataSource
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return articles.count
+        return articlesByCategory[topLevelCategories[section]]!.count
```

Comment:
Use an optional here instead of a ! to unwrap 
`articlesByCategory[topLevelCategories[section]]?.count ?? 0`

or something of the sort so we don't crash just because we have no articles

Reply:
I could do this here pretty cleanly, but I don't want to deal with this case in cellForRowAt and didSelectRowAt.

My argument is that this is a good place to use the `!` b/c it is a programmer error and we want this to crash if our assumptions are wrong so we can fix them.

lines 107,108 from above:
```
articlesByCategory = groupArticlesByTopLevelCategoryName(articles: data)
topLevelCategories = Array(articlesByCategory.keys)
```

we know that topLevelCategories only contains keys of articlesByCategory, so we know that the result of a key lookup will not be nil

Reply:
Yeah that's a good call it should never really be nil in the wild. I'm going to approve.